### PR TITLE
fix: golangci-lint and gosec linters

### PR DESCRIPTION
**Pull request by CamiloGarciaLaRotta**
👋🏽 hi team

### What
This PR addresses gosec and golangci-lint failures in the current main branch. 

It also improves the GoDoc by making the example runnable and by reviewing the syntax of the exported docstrings.

### What approach did you choose and why?
I ran gosec and golangci-lint and tackled each linter failure.

<details><summary>gosec ./...</summary>
<p>

```go
$ gosec ./...
[gosec] 2021/12/10 10:37:45 Including rules: default
[gosec] 2021/12/10 10:37:45 Excluding rules: default
[gosec] 2021/12/10 10:37:45 Import directory: /Users/camilogarcialarotta/playground/srp
[gosec] 2021/12/10 10:37:45 Checking package: srp
[gosec] 2021/12/10 10:37:45 Checking file: /Users/camilogarcialarotta/playground/srp/doc.go
[gosec] 2021/12/10 10:37:45 Checking file: /Users/camilogarcialarotta/playground/srp/group.go
[gosec] 2021/12/10 10:37:45 Checking file: /Users/camilogarcialarotta/playground/srp/internal.go
[gosec] 2021/12/10 10:37:45 Checking file: /Users/camilogarcialarotta/playground/srp/kdf.go
[gosec] 2021/12/10 10:37:45 Checking file: /Users/camilogarcialarotta/playground/srp/proof.go
[gosec] 2021/12/10 10:37:45 Checking file: /Users/camilogarcialarotta/playground/srp/srp.go
[gosec] 2021/12/10 10:37:45 Checking file: /Users/camilogarcialarotta/playground/srp/testonlyexternal.go
[gosec] 2021/12/10 10:37:45 Checking file: /Users/camilogarcialarotta/playground/srp/util.go
Results:


[/Users/camilogarcialarotta/playground/srp/srp.go:161] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
    160:        if s.ephemeralPublicA.Cmp(bigZero) == 0 {
  &gt; 161:                s.makeA()
    162:        }



[/Users/camilogarcialarotta/playground/srp/srp.go:156] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
    155:                if s.ephemeralPublicB.Cmp(bigZero) == 0 {
  &gt; 156:                        s.makeB()
    157:                }



[/Users/camilogarcialarotta/playground/srp/srp.go:143] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
    142:        } else {
  &gt; 143:                s.makeA()
    144:        }



[/Users/camilogarcialarotta/playground/srp/srp.go:141] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
    140:        if s.isServer {
  &gt; 141:                s.makeB()
    142:        } else {



[/Users/camilogarcialarotta/playground/srp/srp.go:137] - G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
    136:        } else {
  &gt; 137:                s.makeLittleK()
    138:        }



Summary:
  Gosec  : dev
  Files  : 8
  Lines  : 1123
  Nosec  : 7
  Issues : 5
```

</p>
</details> 

<details><summary>golangci-lint run</summary>
<p>

```go
$ golangci-lint run
srp_test.go:24:5: `a` is unused (deadcode)
var a = hexNumberString("60975527 035CF2AD 1989806F 0407210B C81EDC04 E2762A56 AFD529DD DA2D4393")
    ^
srp_test.go:36:5: `expectedA` is unused (deadcode)
var expectedA = hexNumberString(
    ^
example_sharedkey_test.go:40:11: Error return value of `rand.Read` is not checked (errcheck)
        rand.Read(salt)
                 ^
example_sharedkey_test.go:171:11: Error return value of `rand.Read` is not checked (errcheck)
        rand.Read(nonce)
                 ^
example_sharedkey_test.go:197:11: Error return value of `rand.Read` is not checked (errcheck)
        rand.Read(nonce) // get a fresh nonce
                 ^
srp.go:137:16: Error return value of `s.makeLittleK` is not checked (errcheck)
                s.makeLittleK()
                             ^
srp.go:141:10: Error return value of `s.makeB` is not checked (errcheck)
                s.makeB()
                       ^
srp.go:143:10: Error return value of `s.makeA` is not checked (errcheck)
                s.makeA()
                       ^
srp.go:156:11: Error return value of `s.makeB` is not checked (errcheck)
                        s.makeB()
                               ^
srp.go:161:10: Error return value of `s.makeA` is not checked (errcheck)
                s.makeA()
                       ^
srp_test.go:71:14: Error return value of `client.makeA` is not checked (errcheck)
        client.makeA()
                    ^
srp_test.go:72:24: Error return value of `client.SetOthersPublic` is not checked (errcheck)
        client.SetOthersPublic(B)
                              ^
srp_test.go:195:24: Error return value of `server.SetOthersPublic` is not checked (errcheck)
        server.SetOthersPublic(A)
                              ^
srp_test.go:272:24: Error return value of `server.SetOthersPublic` is not checked (errcheck)
        server.SetOthersPublic(A)
                              ^
srp_test.go:273:24: Error return value of `client.SetOthersPublic` is not checked (errcheck)
        client.SetOthersPublic(B)
                              ^
```

</p>
</details> 

As an example of the GoDoc changes, left hand side is the current GoDoc, right hand side is the proposed one:
image

### What should reviewers focus on?
The only production code changes are in `srp.go`. I'd like 👀 on it to sanity check that there are no breaking changes in behaviour.

### If something goes wrong, what are the mitigation strategies?
I ran the linters and the test suite locally. They pass. But I see we have no CI, so maybe there are public projects in which I could vendor this latest SHA to sanity check?

### Performance Impact
None

### Observability 
N/A

### CODE REVIEW

**Comment by CamiloGarciaLaRotta** [comment]
FWIW I also started the a PR to introduce basic CI to the repo: https://github.com/CamiloGarciaLaRotta/srp/pull/1

I haven't opened a PR in here, as I want to address and feedback on this one beforehand

File: example_sharedkey_test.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package srp
+package srp_test
```
Comment (jpgoldberg) [inline example_sharedkey_test.go:1]: Ah. That is the right way to organize this. Thank you.

File: example_sharedkey_test.go [hunk 2]
```diff
@@ -37,22 +39,26 @@ func Example() {
 	// Generate 8 bytes of random salt. Be sure to use crypto/rand for all
 	// of your random number needs
 	salt := make([]byte, 8)
-	rand.Read(salt)
+	if n, err := rand.Read(salt); err != nil {
+		log.Fatal(err)
+	} else if n != 8 {
+		log.Fatal("failed to generate 8 byte salt")
+	}
```
Comment (jpgoldberg) [inline example_sharedkey_test.go:47]: Again, thank you. Usually I am the one nagging people to check for errors on reading from the RNG.

File: srp.go [hunk 3]
```diff
@@ -74,13 +74,10 @@ var bigOne = big.NewInt(1)
...
 NewSRPClient sets up an SRP object for a client.
 
-group *Group: Pointer to the Diffie-Hellman group to be used.
-
-x *big.Int: Your long term secret, x.
-
-k *big.Int: If you wish to manually set the multiplier, little k, pass in
-a non-nil bigInt. If you set this to nil, then we will generate one for you.
-You need the same k on both server and client.
+Recall that group is the De-Hellman group to be used,
```
Comment (jpgoldberg) [inline srp.go:77]: s/De-Hellman/Diffie-Hellman/

File: srp.go [hunk 4]
```diff
@@ -134,40 +128,58 @@ func newSRP(serverSide bool, group *Group, xORv *big.Int, k *big.Int) *SRP {
 		// should probably do some sanity checks on k here
 		s.k.Set(k)
 	} else {
-		s.makeLittleK()
+		if _, err := s.makeLittleK(); err != nil {
+			return nil
+		}
```
Comment (jpgoldberg) [inline srp.go:133]: Again, thank you. It was remiss of me to not check errors here.

**Review by jpgoldberg — APPROVED** [review]
Thank you very much for the clean up and just better coding practice all around.

**Pull request merged**